### PR TITLE
Batch restore

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve? 

Scatter region does not work properly when restoring terabyte backup. This leads to an imbalanced cluster.

image

### What is changed and how it works?

Split restore tasks into small batches (128 by default).

image

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: pkg/task/common.go [hunk 1]
```diff
@@ -78,9 +78,14 @@ func DefineCommonFlags(flags *pflag.FlagSet) {
...
 
 	flags.Uint64(flagRateLimit, 0, "The rate limit of the task, MB/s per node")
-	flags.Uint32(flagConcurrency, 4, "The size of thread pool on each node that executes the task")
 	flags.Bool(flagChecksum, true, "Run checksum at end of task")
 
+	// Default concurrency is different for backup and restore.
+	// Leave it 0 and let them adjust the value.
+	flags.Uint32(flagConcurrency, 0, "The size of thread pool on each node that executes the task")
+	// It may confuse users , so just hide it.
+	_ = flags.MarkHidden(flagConcurrency)
```
Comment (kennytm) [inline pkg/task/common.go]: 🤔 but i don't find this confusing

File: pkg/summary/collector_test.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+	c.Assert(len(fields), Equals, 3)
+	assertContains := func(field zap.Field) {
+		for _, f := range fields {
+			if f.Key == field.Key {
+				c.Assert(f, DeepEquals, field)
+				return
+			}
+		}
+		c.Assert(fields, DeepEquals, field) // Must fail and prints detail
```
Comment (kennytm) [inline pkg/summary/collector_test.go]: ??? use [`c.Error`](https://godoc.org/github.com/pingcap/check#C.Error) if this must fail

File: pkg/summary/collector.go [hunk 3]
```diff
@@ -97,19 +104,30 @@ func (tc *logCollector) CollectFailureUnit(name string, reason error) {
 func (tc *logCollector) CollectDuration(name string, t time.Duration) {
 	tc.mu.Lock()
 	defer tc.mu.Unlock()
-	tc.fields = append(tc.fields, zap.Duration(name, t))
+	val, ok := tc.durations[name]
+	if ok {
+		tc.durations[name] = val + t
+		return
+	}
+	tc.durations[name] = t
```
Comment (kennytm) [inline pkg/summary/collector.go]: This entire block can be simplified to

```go
tc.durations[name] += t
```

Demo: https://play.golang.org/p/5s8j5x7GSpl

File: pkg/summary/collector.go [hunk 4]
```diff
@@ -97,19 +104,30 @@ func (tc *logCollector) CollectFailureUnit(name string, reason error) {
...
 func (tc *logCollector) CollectInt(name string, t int) {
 	tc.mu.Lock()
 	defer tc.mu.Unlock()
-	tc.fields = append(tc.fields, zap.Int(name, t))
+	val, ok := tc.ints[name]
+	if ok {
+		tc.ints[name] = val + t
+		return
+	}
+	tc.ints[name] = t
```
Comment (kennytm) [inline pkg/summary/collector.go]: (ditto)

File: pkg/backup/range_test.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backup
```
Comment (kennytm) [inline pkg/backup/range_test.go]: This test should be in the `rtree` package, not here.

File: pkg/backup/range.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	"encoding/hex"
+
+	"github.com/google/btree"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/br/pkg/utils/rtree"
+)
+
+func getIncompleteRange(
```
Comment (kennytm) [inline pkg/backup/range.go]: this function seems more suitable to be in the `rtree` package.

File: pkg/task/restore.go [hunk 7]
```diff
@@ -131,33 +147,55 @@ func RunRestore(c context.Context, g glue.Glue, cmdName string, cfg *RestoreConf
...
 			log.Error("reset pd TS failed", zap.Error(err))
 			return err
 		}
 	}
 
-	removedSchedulers, err := restorePreWork(ctx, client, mgr)
-	if err != nil {
-		return err
+	// Restore sst files in batch.
+	batchSize := int(math.Min(maxRestoreBatchSizeLimit /* 256 */, float64(cfg.Concurrency)))
```
Comment (kennytm) [inline pkg/task/restore.go]:
```suggestion
	batchSize := cfg.Concurrency
	if batchSize > maxRestoreBatchSizeLimit {
		batchSize = maxRestoreBatchSizeLimit // 256
	}
```

or use `mathutil.Min(cfg.Concurrency, maxRestoreBatchSizeLimit)` if you must need this to be single-line.

File: pkg/task/restore.go [hunk 8]
```diff
@@ -131,33 +147,55 @@ func RunRestore(c context.Context, g glue.Glue, cmdName string, cfg *RestoreConf
...
+	batches := make([][]rtree.Range, 0, (len(ranges)+batchSize-1)/batchSize)
+	for batchSize < len(ranges) {
+		ranges, batches = ranges[batchSize:], append(batches, ranges[0:batchSize:batchSize])
 	}
-	err = client.RestoreFiles(files, rewriteRules, updateCh)
-	// always run the post-work even on error, so we don't stuck in the import mode or emptied schedulers
-	postErr := restorePostWork(ctx, client, mgr, removedSchedulers)
+	batches = append(batches, ranges)
 
+	for _, rangeBatch := range batches {
```
Comment (kennytm) [inline pkg/task/restore.go]: seems constructing the `batches` array is unnecessary? the two `for` loops can be merged?

File: pkg/task/common.go (hunk 1 above)
Comment (overvenus) [reply pkg/task/common.go]: `--help` prints "concurrency .... (default 0)", I find this is confusing. :joy:

Comment (kennytm) [reply pkg/task/common.go]: `(default 0)` should not appear, otherwise the `--rate-limit` would also show `(default 0)`.

**Comment by 5kbpers** [comment]
Do you know why scatter region does not work? Is it because the timeout is too short?

**Comment by overvenus** [comment]
> Do you know why scatter region does not work? Is it because the timeout is too short?

In short, yes.

In the test, we are scattering ~40k region at once, if a cluster is configured 5 nodes 3 replicas, it means PD needs to scatter ~48k replicas in a single scatter operator.

Based on my test, PD takes 15s to scatter 156 replicas, it is clearly scattering 48k replicas  will be timeout.

File: pkg/rtree/check.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+import (
+	"encoding/hex"
+
+	"github.com/google/btree"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+)
+
+// CheckDupFiles checks if there are any files are duplicated.
+func CheckDupFiles(rangeTree *RangeTree) {
```
Comment (kennytm) [inline pkg/rtree/check.go]: IMO this function is irrelevant to RangeTree and should be placed in its original package. Not a big deal anyway.

**Pull request merged**